test/PushNotificationTests/BaseTestSuite.cpp

```diff
@@ -0,0 +1,264 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+
+#include <TestDef.h>
+#include "BaseTestSuite.h"
+
+using namespace WEX::Common;
+using namespace WEX::Logging;
+using namespace WEX::TestExecution;
+
+using namespace winrt::Windows::ApplicationModel::Activation;
+using namespace winrt::Windows::ApplicationModel::Background;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Foundation::Collections;
+using namespace winrt::Windows::Management::Deployment;
+using namespace winrt::Windows::Storage;
+using namespace winrt::Windows::System;
+using namespace winrt::Microsoft::Windows::PushNotifications;
+
+void BaseTestSuite::ClassSetup()
+{
+    ::Test::Bootstrap::Setup();
+    bool isSelfContained{ false };
+    if (SUCCEEDED(TestData::TryGetValue(L"SelfContained", isSelfContained)))
+    {
+        if (!isSelfContained)
+        {
+            ::WindowsAppRuntime::SelfContained::TestInitialize(::Test::Bootstrap::TP::WindowsAppRuntimeFramework::c_PackageFamilyName);
+        }
+    }
+}
+
+void BaseTestSuite::ClassCleanup()
+{
+    ::Test::Bootstrap::Cleanup();
+}
+
+void BaseTestSuite::MethodCleanup()
+{
+    if (m_registered)
+    {
+        PushNotificationManager::Default().UnregisterAll();
+        m_registered = false;
+    }
+}
+
+HRESULT BaseTestSuite::ChannelRequestHelper(IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& channelOperation)
+{
+    if (channelOperation.wait_for(c_timeout) != winrt::Windows::Foundation::AsyncStatus::Completed)
+    {
+        channelOperation.Cancel();
+        return HRESULT_FROM_WIN32(ERROR_TIMEOUT); // timed out or failed
+    }
+
+    auto result{ channelOperation.GetResults() };
+    auto status{ result.Status() };
+    if (status != PushNotificationChannelStatus::CompletedSuccess)
+    {
+        return result.ExtendedError(); // did not produce a channel
+    }
+
+    result.Channel().Close();
+    return S_OK;
+}
+
+void BaseTestSuite::ChannelRequestUsingNullRemoteId()
+{
+    if (PushNotificationManager::Default().IsSupported())
+    {
+        VERIFY_THROWS_HR(PushNotificationManager::Default().CreateChannelAsync(winrt::guid()).get(), E_INVALIDARG);
+    }
+    else
+    {
+        auto channelOperation{ PushNotificationManager::Default().CreateChannelAsync(winrt::guid()) };
+        VERIFY_ARE_EQUAL(ChannelRequestHelper(channelOperation), E_FAIL);
```

Comment:
`E_FAIL`? Sad now.

I have a deep dislike for E_FAIL. "I had a problem, and yeah, I know what it is, but not gonna tell you nyah nyah".

That's coming out of the actual PushNotification API? You may want to consider a more specific error code in a future change.

P.S. You can invent new HRESULTs if necessary using FACILITY_ITF. There's caveats to FACILITY_ITF so probably not something to go hog wild on, but when valuable we have the option. See MsixDynamicDependency.h for some examples.

Reply:
`HRESULT_FROM_WIN32(ERROR_NOT_ENOUGH_QUOTA)` 🙃 I kid, I kid.

Reply:
The error is a developer attempting to call a PushNotification APIs in SelfContained environment which is unsupported in this specific scenario. What would be a better existing error code to return? Or should we invent a new HRESULT in this scenario?

Reply:
`HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED)` is that what describe

Reply:
@pmpurifoy Lets open a bug to address this.

Reply:
https://github.com/microsoft/WindowsAppSDK/issues/2416